src/Compilers/CSharp/Portable/Compilation/CSharpCompilation.cs

```diff
@@ -1730,6 +1730,18 @@ bool checkValid(MethodSymbol candidate, bool isCandidate, DiagnosticBag specific
                         }
                     }
                 }
+                else if (LanguageVersion >= MessageID.IDS_FeatureAsyncMain.RequiredVersion() && taskEntryPoints.Count > 0)
+                {
+                    // TODO: I think I should first concat both taskEntryPoints and viableEntryPoints first.
+                    viableEntryPoints.Sort(LexicalOrderSymbolComparer.Instance);
+                    var info = new CSDiagnosticInfo(
+                         ErrorCode.WRN_SyncAndAsyncEntryPoints,
+                         args: Array.Empty<object>(),
+                         symbols: viableEntryPoints.OfType<Symbol>().AsImmutable(),
+                         additionalLocations: viableEntryPoints.Select(m => m.Locations.First()).OfType<Location>().AsImmutable());
+
+                    diagnostics.Add(new CSDiagnostic(info, viableEntryPoints.First().Locations.First()));
```

Comment:
Consider extracting this to a static local function, and using here and below.

Reply:
The above logic is likely to get changed a little bit based on the current TODO comment I've. Or are you okay if I ignored this TODO?

Reply:
It still seems like this is common code that can be extracted.

---
In reply to: [472394045](https://github.com/dotnet/roslyn/pull/46832#discussion_r472394045) [](ancestors = 472394045)

Reply:
@333fred Do you mean this whole block:

```

                    var info = new CSDiagnosticInfo(
                         ErrorCode.WRN_SyncAndAsyncEntryPoints,
                         args: Array.Empty<object>(),
                         symbols: ImmutableArray<Symbol>.Create(viableEntryPoints.First(), taskEntryPoints.First()),
                         additionalLocations: ImmutableArray<Location>.Create(viableEntryPoints.First().Locations.First(), taskEntryPoints.First().Locations.First()));

                    diagnostics.Add(new CSDiagnostic(info, viableEntryPoints.First().Locations.First())); 
```

or just the last line of it?

Reply:
I mean the whole block. It seems like we should just be able to pass an array of entry points to the helper, the diagnostics, and a bool for "warn or error", and it can do the rest.

Reply:
This is no longer relevant with the new warning changes.